src/Nethermind/Nethermind.Baseline/MerkleTree.cs

```diff
@@ -0,0 +1,232 @@
+            }
+
+            return Bytes32.Wrap(nodeHashBytes);
+        }
+
+        internal static uint GetIndexAtLevel(uint level, uint nodeIndex)
+        {
+            ValidateLevel(level);
+            ValidateNodeIndex(nodeIndex);
+
+            uint indexAtLevel = nodeIndex - ((1u << (int) level) - 1);
+            ValidateIndexAtLevel(level, indexAtLevel);
+            return indexAtLevel;
+        }
+
+        internal static uint GetLeafIndex(uint nodeIndex)
+        {
+            if (nodeIndex == uint.MaxValue ||
+                nodeIndex < FirstLeafIndexAsNodeIndex)
+            {
+                throw new IndexOutOfRangeException($"Leaf indices start at {FirstLeafIndexAsNodeIndex}");
+            }
+
+            return nodeIndex - FirstLeafIndexAsNodeIndex;
+        }
+
+        internal static uint GetSiblingIndex(uint level, uint indexAtLevel)
+        {
+            ValidateLevel(level);
+            ValidateIndexAtLevel(level, indexAtLevel);
+
+            if (level == 0)
+            {
+                throw new IndexOutOfRangeException("Root node has no siblings.");
+            }
+            
+            if (indexAtLevel % 2 == 0)
+            {
+                return indexAtLevel + 1;
+            }
+
+            return indexAtLevel - 1;
+        }
+
+        internal static void ValidateNodeIndex(uint nodeIndex)
+        {
+            if (nodeIndex > MaxNodeIndex)
+            {
+                throw new IndexOutOfRangeException($"Node index should be between 0 and {MaxNodeIndex}");
+            }
+        }
+
+        internal static void ValidateLevel(uint level)
+        {
+            if (level > LeafLevel)
+            {
+                throw new IndexOutOfRangeException($"Tree level should be between 0 and {LeafLevel}");
+            }
+        }
+
+        internal static void ValidateIndexAtLevel(uint level, uint indexAtLevel)
+        {
+            uint maxIndexAtLevel = (1u << (int) level) - 1u;
+            if (indexAtLevel > maxIndexAtLevel)
+            {
+                throw new IndexOutOfRangeException($"Tree level {level} should only has indices between 0 and {maxIndexAtLevel}");
+            }
+        }
+
+        internal static uint GetNodeIndex(uint level, uint indexAtLevel)
+        {
+            ValidateLevel(level);
+            ValidateIndexAtLevel(level, indexAtLevel);
+
+            return (1u << (int) level) - 1u + indexAtLevel;
+        }
+
+        internal static uint GetParentIndex(uint nodeIndex)
+        {
+            ValidateNodeIndex(nodeIndex);
+
+            if (nodeIndex == 0)
+            {
+                throw new IndexOutOfRangeException("Root node has no parent");
+            }
+
+            return (nodeIndex + 1) / 2 - 1;
+        }
+
+        [MethodImpl(MethodImplOptions.Synchronized)]
+        public void Insert(Bytes32 leaf)
+        {
+            _keyValueStore[Rlp.Encode(Count).Bytes] = leaf.AsSpan().ToArray();
```

Comment:
Can we have Get/Set extension methods for encoding keys?

Reply:
it is wrapped to Save and Load methods in the BaselineTree - MerkleTree will be moved out of here to Eth2